src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+    True
+    >>> _overlaps_one_side(range(8, 12), range(4, 8))  # fully to right
+    False
+    """
+    assert a.step == b.step and a.step in (1, None)
+    return (a.start < b.start < a.stop < b.stop or
+            a.start == b.start < a.stop < b.stop or
+            b.start < a.start < b.stop < a.stop)
+
+
+@dataclass
+class Cursor(ABC):
+    _proc: ReferenceType[ProcedureBase]
+
+    # ------------------------------------------------------------------------ #
+    # Static constructors
+    # ------------------------------------------------------------------------ #
+
+    @staticmethod
+    def root(proc: ProcedureBase):
+        return Node(weakref.ref(proc), [])
+
+    # ------------------------------------------------------------------------ #
+    # Validating accessors
+    # ------------------------------------------------------------------------ #
+
+    def proc(self):
+        if (p := self._proc()) is None:
+            raise InvalidCursorError("underlying proc was destroyed")
+        return p
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (abstract)
+    # ------------------------------------------------------------------------ #
+
+    @abstractmethod
+    def parent(self) -> Node:
+        """Get the node containing the current cursor"""
+
+    @abstractmethod
+    def before(self, dist=1) -> Cursor:
+        """For gaps, get the node before the gap. Otherwise, get the gap before
+        the node or block"""
+
+    @abstractmethod
+    def after(self, dist=1) -> Cursor:
+        """For gaps, get the node after the gap. Otherwise, get the gap after
+        the node or block"""
+
+    @abstractmethod
+    def prev(self, dist=1) -> Cursor:
+        """Get the previous node/gap in the block. Undefined for blocks."""
+
+    @abstractmethod
+    def next(self, dist=1) -> Cursor:
+        """Get the next node/gap in the block. Undefined for blocks."""
+
+    # ------------------------------------------------------------------------ #
+    # Protected path / mutation helpers
+    # ------------------------------------------------------------------------ #
+
+    def _translate(self, ty, path, dist):
+        if not path:
+            raise InvalidCursorError('cannot move root cursor')
+        attr, i = path[-1]
+        if i is None:
+            raise InvalidCursorError('cursor is not inside block')
+        return ty(self._proc, path[:-1] + [(attr, i + dist)])
+
+    @staticmethod
+    def _walk_path(n, path):
+        for attr, idx in path:
+            n = getattr(n, attr)
+            if idx is not None:
+                n = n[idx]
+        return n
+
+    def _rewrite_node(self, fn):
+        """
+        Applies `fn` to the AST node containing the cursor. The callback gets
+        the following arguments:
+        1. The parent node
+        2. The relevant list of children
+        3. The last step in the path (the attribute and index info)
```

Comment:
This seems to not be true?  Below `fn()` only gets a single node.  (The parent?)

Reply:
Ah, yes, this comment is stale. I removed the latter two arguments after I found they weren't useful. Only the parent node is passed. Thanks for catching this.